rfcs/2022-10-12-14742-flexible-metric-tags.md

```diff
@@ -0,0 +1,191 @@
+
+## Scope
+
+### In scope
+
+- Changes to the existing metric tag internal and external representation to support multiple values
+  and bare tags.
+
+### Out of scope
+
+- Changes to any other aspect of metric or event representations.
+
+## Pain
+
+When ingesting data from a Datadog agent, the metric tags provided by that source may contain data
+that Vector's data model cannot represent. This is causing problems for prospective users that want
+to add Vector to their observability pipelines but cannot do so because it would cause data loss.
+
+## Proposal
+
+### User Experience
+
+This change should be effectively invisible to users of Vector, except for incoming data with
+multiple tags with the same name, which will now have all the tags reproduced when it reaches sinks.
+
+### Implementation
+
+#### Internal Representation
+
+Simply put, the tags representation will change from an alias into a newtype wrapper using an
+`indexmap` set to store the tag values. This newtype will hide the implementation details of the
+underlying storage from the callers. It will also add separate methods for inserting a new tag and
+replacing a tag set with a single value. The callers of the existing `insert` function will need to
+be audited to determine which use is intended at each call site.
+
+```rust
+type TagValue = Option<String>;
+
+struct MetricTags(BTreeMap<String, indexmap::IndexSet<TagValue>>);
+
+impl MetricTags {
+    // Insert returns the value unchanged if the exact (tag,value) pair already exists,
+    // otherwise it inserts a new value for the named tag.
+    fn insert(&mut self, name: String, value: TagValue) -> Option<TagValue>;
+
+    // Replace returns all the existing values when overwriting a tag.
+    fn replace(&mut self, name: String, value: Option<TagValue>) -> Option<IndexSet<TagValue>>;
+}
+```
+
+#### External Representation
+
+The existing Protobuf representation of metric tags has a map of string-to-string pairs of
+tags. A new tags map will be added where the values are lists of optional strings, matching the
+semantics given above.
+
+```protobuf
+message Metric {
+  …
+  map<string, string> tags_v1 = 3;
+  message Values {
+    message Value {
+      optional string value = 1;
+    }
+    repeated Value values = 1;
+  }
+  map<string, Values> tags_v2 = 20;
+  …
+}
+```
+
+## Rationale
+
+Changing the `MetricTags` type from an alias to a newtype wrapper allows us to provide better
+compatibility for existing uses while controlling the methods for uses that need to access all the
+values.
+
+The use of an `IndexSet` for the tag value provides us with two useful invariants:
+
+1. Only unique values for each tag will be stored, which prevents repeated values from showing up in
+   the output.
+1. The values can be retrieved in the order they first appeared, which allows us to trivially
+   retrieve either the first or last stored value.
```

Comment:
Given the structuring, it seems like we'll only preserve ordering within a given tag key itself. Do we have any inkling around whether or not total ordering (considering all tag keys) is required/expected/present in other products?

Reply:
I am not aware of any ordering requirement in other products. My understanding is that the Datadog programs reorder tags internally to handle consistent hashing, so we should be free to order them however is convenient for us.